deepchem/models/tensorgraph/graph_layers.py

```diff
@@ -799,3 +841,258 @@ def DAGgraph_step(self, batch_inputs, W_list, b_list):
+
+  def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
+    """ Perform T steps of message passing """
+    if in_layers is None:
+      in_layers = self.in_layers
+    in_layers = convert_to_layers(in_layers)
+
+    # Extract atom_features
+    atom_features = in_layers[0].out_tensor
+    pair_features = in_layers[1].out_tensor
+    atom_to_pair = in_layers[2].out_tensor
+    n_atom_features = atom_features.get_shape().as_list()[-1]
+    n_pair_features = pair_features.get_shape().as_list()[-1]
+    # Add trainable weights
+    self.build(pair_features, n_pair_features)
+
+    if n_atom_features < self.n_hidden:
+      pad_length = self.n_hidden - n_atom_features
+      out = tf.pad(atom_features, ((0, 0), (0, pad_length)), mode='CONSTANT')
+    elif n_atom_features > self.n_hidden:
+      raise ValueError("Too large initial feature vector")
+
+    for i in range(self.T):
+      message = self.message_function.forward(out, atom_to_pair)
+      out = self.update_function.forward(out, message)
+
+    out_tensor = out
+
+    if set_tensors:
+      self.variables = self.trainable_weights
+      self.out_tensor = out_tensor
+    return out_tensor
+
+  def none_tensors(self):
+    message_tensors = self.message_function.none_tensors()
+    update_tensors = self.update_function.none_tensors()
+    out_tensor, trainable_weights, variables = self.out_tensor, self.trainable_weights, self.variables
+    self.out_tensor, self.trainable_weights, self.variables = None, [], []
+    return message_tensors, update_tensors, out_tensor, trainable_weights, variables
+
+  def set_tensors(self, tensor):
+    message_tensors, update_tensors, self.out_tensor, self.trainable_weights, self.variables = tensor
+    self.message_function.set_tensors(message_tensors)
+    self.update_function.set_tensors(update_tensors)
+
+
+class EdgeNetwork(object):
+  """ Submodule for Message Passing """
+
+  def __init__(self,
+               pair_features,
+               n_pair_features=8,
+               n_hidden=100,
+               init='glorot_uniform'):
+    self.n_pair_features = n_pair_features
+    self.n_hidden = n_hidden
+    self.init = initializations.get(init)
+    W = self.init([n_pair_features, n_hidden * n_hidden])
+    b = model_ops.zeros(shape=(n_hidden * n_hidden,))
+    self.A = tf.nn.xw_plus_b(pair_features, W, b)
+    self.A = tf.reshape(self.A, (-1, n_hidden, n_hidden))
+    self.trainable_weights = [W, b]
+
+  def forward(self, atom_features, atom_to_pair):
+    out = tf.expand_dims(tf.gather(atom_features, atom_to_pair[:, 1]), 2)
+    out = tf.reduce_sum(out * self.A, axis=1)
+    out = tf.segment_sum(out, atom_to_pair[:, 0])
+    return out
+
+  def none_tensors(self):
+    A = self.A
+    self.A = None,
+    trainable_weights = self.trainable_weights
+    self.trainable_weights = []
+    return A, trainable_weights
+
+  def set_tensors(self, tensor):
+    self.A, self.trainable_weights = tensor
+
+
+class GatedRecurrentUnit(object):
```

Comment:
Is this the same as https://github.com/deepchem/deepchem/blob/master/deepchem/models/tensorgraph/layers.py#L401, the existing GRU class? Could we replace it and use the existing GRU?